### PR TITLE
Add streaming microphone sample for Speech

### PR DESCRIPTION
**Pull request by CallistoCF**
For @beccca preliminary review

Command line microphone streaming example in node, run MicrophoneStream.js --help

### CODE REVIEW

File: samples/system-test/microphonestream.test.js [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+//const uuid = require(`uuid`);
+
+const {runAsync} = require(`@google-cloud/nodejs-repo-tools`);
+
+//const bucketName = `nodejs-docs-samples-test-${uuid.v4()}`;
+const cmd = `node MicrophoneStream.js`;
+const cwd = path.join(__dirname, `..`);
+const filename = `ManualTest.wav`;
+const filepath = path.join(__dirname, `../resources/${filename}`);
+const text = `neewer manual test 10:39 Friday`;
```
Comment (beccca) [inline samples/system-test/microphonestream.test.js]: Where does this text come from? Is `neewer` a word?

File: samples/MicrophoneStream.js [hunk 2]
```diff
@@ -0,0 +1,377 @@
...
+      const response = data[0];
+      const transcription = response.results
+        .map(result => result.alternatives[0].transcript)
+        .join('\n');
+      console.log(`wavFileToText: Transcription; \n`, transcription);
+      const confidence = response.results
+        .map(result => result.alternatives[0].confidence)
+        .join('\n');
+      console.log(`wavFileToText: Confidence; \n`, confidence);
+      // ResultsArray.push([transcription, confidence]);
```
Comment (beccca) [inline samples/MicrophoneStream.js]: Is this old code? What's up with this commented out code?

File: samples/system-test/microphonestream.test.js [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+'use strict';
+
+const path = require(`path`);
+//const storage = require(`@google-cloud/storage`)();
+const test = require(`ava`);
+//const uuid = require(`uuid`);
+
+const {runAsync} = require(`@google-cloud/nodejs-repo-tools`);
+
+//const bucketName = `nodejs-docs-samples-test-${uuid.v4()}`;
```
Comment (beccca) [inline samples/system-test/microphonestream.test.js]: Why is this commented out?

File: samples/system-test/microphonestream.test.js [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+'use strict';
+
+const path = require(`path`);
+//const storage = require(`@google-cloud/storage`)();
```
Comment (beccca) [inline samples/system-test/microphonestream.test.js]: There are multiple commented out variables here and throughout the file. Could you provide some insight as to why? Are these left over from testing/playing you were doing? Can I expect these to be changed or removed?

File: samples/system-test/microphonestream.test.js [hunk 5]
```diff
@@ -0,0 +1,45 @@
+/**
+ * Copyright 2016, Google, Inc.
```
Comment (beccca) [inline samples/system-test/microphonestream.test.js]: Nit: update licenses to be `2018` _( i realize you copy/pasted from an older file, which is what I always do, we just like to use the current date on new files )_

File: samples/MicStream.js [hunk 6]
```diff
@@ -0,0 +1,332 @@
+/**
```
Comment (beccca) [inline samples/MicStream.js]: I see 2 different .js files for MicStreaming MicrophoneStreaming – should there be just 1? Which one is up-to-date? Will the later be removed?

Comment (CallistoCF) [reply samples/MicStream.js]: Just pushed commit d13734f - my bad, i included earlier draft - everything should be in micstream.js - running 'node micstream.js --help' lists yaaaarghs

File: samples/MicStream.js [hunk 7]
```diff
@@ -0,0 +1,378 @@
...
+  }
+  console.log(`stream speech to text is: ${stream}`);
+  console.log(`   wav speech to text is: ${wav}`);
+}
+
+require(`yargs`)
+  .demand(1)
+  .command(
+    `run`,
+    `Runs complete streaming test: \n 1. streams from microphone, outputting streamed speech-to-text to console, \n 2. records that streamed sample to resources folder, \n 3. runs speech-to-text on that local file to compare to step 1.`,
```
Comment (beccca) [inline samples/MicStream.js]: Cool idea, but not part of this sample

Let's remove the code that saves to a .wav file, it's out of scope (and it's how people usually use Speech-to-Text, see the existing samples in this directory)

**Comment by beccca** [comment]
Where are the tests?

File: samples/MicStream.js [hunk 8]
```diff
@@ -0,0 +1,378 @@
...
+
+  console.log(
+    `micStreamRecognize: Listening for ${secondsofRecording} seconds, press Ctrl+C to stop.`
+  );
+
+  //CF: records for secondsofRecording seconds.
+  setTimeout(function() {
+    record.stop();
+    console.log(`micStreamRecognize: Recording complete!`);
+    console.log(
```
Comment (beccca) [inline samples/MicStream.js]: The results aren't printing while the process is running, I have to wait until the end to see results

Please update to print results as they come in

File: samples/MicStream.js [hunk 9]
```diff
@@ -0,0 +1,378 @@
...
+    for (var i in InterimResults) {
+      console.log(`  `, i, `)`, InterimResults[i].toString());
+    }
+    eventEmitter.emit(`micStreamRecognize: complete`);
+  }, secondsofRecording * 1000);
+  return InterimResults;
+  // [END micStreamRecognize]
+}
+
+function wavFileToText(filename, encoding, sampleRateHertz, languageCode) {
```
Comment (beccca) [inline samples/MicStream.js]: We already have a samples for this:
 - https://cloud.google.com/speech-to-text/docs/sync-recognize#speech-sync-recognize-nodejs
 - https://github.com/googleapis/nodejs-speech/blob/master/samples/recognize.js#L27-L70

Remove from the mic streaming sample code

File: samples/MicStream.js [hunk 10]
```diff
@@ -0,0 +1,378 @@
...
+      console.log(`wavFileToText: Confidence; \n`, confidence);
+      eventEmitter.emit(`wavFileToText: complete`);
+    })
+    .catch(err => {
+      console.error('ERROR:', err);
+    });
+  // [END wavFileToText]
+}
+
+function test_1(
```
Comment (beccca) [inline samples/MicStream.js]: Please move tests out of this file & into the system-tests

e.g. how about this:

`microphoneStreaming.js` & `system-tests/microphoneStreaming.test.js`

File: samples/MicStream.js [hunk 11]
```diff
@@ -0,0 +1,378 @@
...
+  const speech = require('@google-cloud/speech');
+
+  // Creates a client
+  const client = new speech.SpeechClient();
+
+  //Creates an array of interim results, for testing
+  var InterimResults = [];
+
+  //defensive code in case user doesn't specify audio file type
+  if (filename.substr(filename.length - 4) !== `.wav`)
```
Comment (beccca) [inline samples/MicStream.js]: Move this into a default value for the CLI params

File: samples/MicStream.js [hunk 12]
```diff
@@ -0,0 +1,378 @@
...
+  const record = require('node-record-lpcm16');
+  //https://www.npmjs.com/package/node-record-lpcm16
+
+  // Imports the Google Cloud client library
+  const speech = require('@google-cloud/speech');
+
+  // Creates a client
+  const client = new speech.SpeechClient();
+
+  //Creates an array of interim results, for testing
```
Comment (beccca) [inline samples/MicStream.js]: Remove all test-related code from the sample file

File: samples/MicStream.js [hunk 13]
```diff
@@ -0,0 +1,378 @@
...
+    config,
+    interimResults: true, // If you want interim results, set this to true
+  };
+
+  //To write .wav to file
+  const fs = require('fs');
+
+  //Node-Record-lpcm16
+  const record = require('node-record-lpcm16');
+  //https://www.npmjs.com/package/node-record-lpcm16
```
Comment (beccca) [inline samples/MicStream.js]: Remove URL

File: samples/MicStream.js [hunk 14]
```diff
@@ -0,0 +1,378 @@
...
+    sampleRateHertz: sampleRateHertz,
+    languageCode: languageCode,
+  };
+
+  const request = {
+    config,
+    interimResults: true, // If you want interim results, set this to true
+  };
+
+  //To write .wav to file
```
Comment (beccca) [inline samples/MicStream.js]: Unnecessary, remove

File: samples/MicStream.js [hunk 15]
```diff
@@ -0,0 +1,378 @@
...
+
+  const config = {
+    encoding: encoding,
+    sampleRateHertz: sampleRateHertz,
+    languageCode: languageCode,
+  };
+
+  const request = {
+    config,
+    interimResults: true, // If you want interim results, set this to true
```
Comment (beccca) [inline samples/MicStream.js]: How about...

```js
interimResults: true, // Get interim results from stream
```

File: samples/MicStream.js [hunk 16]
```diff
@@ -0,0 +1,378 @@
...
+
+function micStreamRecognize(
+  filename,
+  secondsofRecording,
+  encoding,
+  sampleRateHertz,
+  languageCode
+) {
+  // [START micStreamRecognize]
+  //returns speech-to-text, from streamingRecognize - request object is simplified
```
Comment (beccca) [inline samples/MicStream.js]: remove comment

File: samples/MicStream.js [hunk 17]
```diff
@@ -0,0 +1,378 @@
...
+  //streams audio from microphone to filename, for secondsofRecording seconds, using simplified request object
+  //config is .wav by default, change request object (see above) for more options.
```
Comment (beccca) [inline samples/MicStream.js]: remove comment

File: samples/MicStream.js [hunk 18]
```diff
@@ -0,0 +1,378 @@
...
+function micStreamRecognize(
+  filename,
+  secondsofRecording,
+  encoding,
+  sampleRateHertz,
+  languageCode
+) {
+  // [START micStreamRecognize]
+  //returns speech-to-text, from streamingRecognize - request object is simplified
+  //streams audio from microphone to filename, for secondsofRecording seconds, using simplified request object
```
Comment (beccca) [inline samples/MicStream.js]: remove comment

File: samples/MicStream.js [hunk 19]
```diff
@@ -0,0 +1,378 @@
...
+ * 1. SoX must be installed and available in your $PATH- it can be found here:
+ * http://sox.sourceforge.net/
+ * 2. Microphone must be working
+ * 3. Encoding, sampleRateHertz, and # of channels must match header of audio file you're
+ * recording to.
+ * 4. Get Node-Record-lpcm16 https://www.npmjs.com/package/node-record-lpcm16
+ * More Info: https://cloud.google.com/speech-to-text/docs/streaming-recognize
+ */
+
+//uncomment below and see Errata at end of file for more configurations
```
Comment (beccca) [inline samples/MicStream.js]: We only use uncommented "placeholder variables" like this inside of `[START` `END]` "region tags" for embedding in the documentation

I'll follow up privately with a guide for sample style

File: samples/MicrophoneStream.js [hunk 20]
```diff
@@ -0,0 +1,129 @@
+/**
+ * Copyright 2017, Google, Inc.
```
Comment (happyhuman) [inline samples/MicrophoneStream.js]: 2018 LLC

File: samples/system-test/MicrophoneStream.test.js [hunk 21]
```diff
@@ -0,0 +1,30 @@
+/**
+ * Copyright 2016, Google, Inc.
```
Comment (happyhuman) [inline samples/system-test/MicrophoneStream.test.js]: 2018 LLC

**Comment by JustinBeckwith** [comment]
@beccca @happyhuman 👋 What's the status of this PR?  It's been hanging around for a while :) Anything we can do to help get it merged?

File: README.md [hunk 22]
```diff
@@ -2,149 +2,140 @@
 [//]: # "To regenerate it, use `npm run generate-scaffolding`."
 <img src="https://avatars2.githubusercontent.com/u/2810941?v=3&s=96" alt="Google Cloud Platform logo" title="Google Cloud Platform" align="right" height="96" width="96"/>
 
-# [Google Cloud Speech API: Node.js Client](https://github.com/googleapis/nodejs-speech)
+# Google Cloud Speech API: Node.js Samples
```
Comment (stephenplusplus) [inline README.md]: @JustinBeckwith this PR replaced the root README with the samples/README.md. Was that intended?

Comment (JustinBeckwith) [reply README.md]: 🤦‍♂️  No, I don't think that was the intent.  Nice catch @stephenplusplus.  Filed #171 to track.

Comment (beccca) [reply README.md]: Oh, crud. I should've caught this. Thanks @stephenplusplus!

**Pull request merged**